.github/workflows/cut-rc.yml

```diff
@@ -0,0 +1,503 @@
+        run: |
+          git config user.name "${{github.event.sender.login}}"
+          git config user.email noreply@github.com
+      - run: |
+          # Could be optimized: done only if there are closed PRs to cherry-pick
+          git fetch origin ${{env.main_branch}}:${{env.main_branch}} --unshallow
+      - env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        name: Cherry-pick commits from the main branch
+        id: cherry-pick
+        run: |
+          SLACK_MESSAGE_FILE=$(mktemp)
+          EMPTY=""
+          gh pr list -s closed \
+            --search "milestone:${{needs.variables.outputs.milestone}}" \
+            --base "${{env.main_branch}}" \
+            --json number,url,mergeCommit,author,title \
+            --jq '.[] | "\(.number)\t\(.url)\t\(.mergeCommit.oid)\t\(.author.login)\t\(.title)"' \
+          | while read -r PR_COMMIT
+          do
+            IFS=$'\t' read -r PR URL COMMIT AUTHOR TITLE <<< $PR_COMMIT
+            # Skip commits merged before branching.
+            if git merge-base --is-ancestor "$COMMIT" HEAD
+            then
+              continue
+            fi
+
+            FORK_POINT=$(diff -u \
+              <(git rev-list --first-parent ${{needs.variables.outputs.branch}}) \
+              <(git rev-list --first-parent ${{env.main_branch}}) \
+              | sed -ne 's/^ //p' | head -1)
+            # Find commits with the specific commit message before the fork point
+            if [[ -z "$(git log \
+              --grep "^(cherry picked from commit $COMMIT)\$" --format='%H' \
+              HEAD...$FORK_POINT )" ]]
+            then
+              if git cherry-pick -x "$COMMIT"
+              then
+                echo "Cherry-picked $COMMIT from PR $PR" >> $GITHUB_STEP_SUMMARY
+                [ "${{github.event.inputs.dry-run}}" != "true" ] && \
+                gh pr comment $PR --body "Merge commit has been cherry-picked to branch \`${{needs.variables.outputs.branch}}\`."
+                EMPTY=false
+              else
+                git cherry-pick --abort
+                [ "${{github.event.inputs.dry-run}}" != "true" ] && \
+                gh pr comment $PR --body "Please merge the changes to branch \`${{needs.variables.outputs.branch}}\`."
+                cat << EOF >> $GITHUB_STEP_SUMMARY
+          * [PR $PR]($URL) by **${AUTHOR}** (_${TITLE}_) could not be cherry-picked. Merge commit: \`$COMMIT\`.
+          EOF
+                echo "- <$URL|PR $PR> by *$AUTHOR* — $TITLE" \
+                >> "$SLACK_MESSAGE_FILE"
+              fi
+            fi
+          done
+
+          git push
+          if [ -n "$EMPTY" ]; then
+            echo "Cherry-picked commits have been pushed to the release branch." >> $GITHUB_STEP_SUMMARY
+          fi
+
+          # Replace % with %25 and join lines with %0A:
+          FAILED="$(sed ':a; N; $!ba; s/%/%25/g; s/\n/%0A/g' $SLACK_MESSAGE_FILE)"
+          rm -f "$SLACK_MESSAGE_FILE"
+
+          if [ -n "$FAILED" ]
```

Comment:
I do not understand how the failed sed on a slack-message file is used to determine whether cherry-picking has succeeded. Could we think of providing some more robust check?

Reply:
SLACK_MESSAGE_FILE file is basically a list of failed PRs to be reported to SLACK. /* I could probably use bash arrays, but I avoid advanced bash features as you could have noticed */
So if it is empty, there were no merge failures. I considered it to be a good check.